### ISSUE TITLE
Throwing UnauthorizedResponse when redirecting to login page

### ISSUE THREAD

**Issue by jehugaleahsa** [pac4j/javalin-pac4j#129]
In `SecurityHandler`, there is the following code:

```java
if (result != "AUTH_GRANTED") {
    throw new UnauthorizedResponse();
}
```

When navigating to a page in the browser, you typically want to be redirected to the login page via 302. When hitting an API end-point, you typically want to receive a 401. There's code built into PAC4J that detects whether it is an AJAX call or not and takes the appropriate action. This check prevents this out-of-the-box behavior provided by PAC4J.

I basically stole the code from inside `SecurityHandler` and made something as simple as this:
```java
        SessionStore sessionStore = FindBest.sessionStore(null, this.config, JEESessionStore.INSTANCE);
        HttpActionAdapter adapter = FindBest.httpActionAdapter(null, this.config, JavalinHttpActionAdapter.INSTANCE);
        SecurityLogic bestLogic = FindBest.securityLogic(null, this.config, DefaultSecurityLogic.INSTANCE);

        JavalinWebContext context = new JavalinWebContext((Context) httpContext.getHandle());
        Object result = bestLogic.perform(
            context,
            sessionStore,
            this.config,
            (ctx, store, profiles, parameters) -> new Object(),
            adapter,
            "FormClient",
            null,
            null);
        return result != null;
```

I am just hard-coding the last three arguments in my example. I am okay using my code for now, but thought I should report this since it seems like it might impact other people.

**Comment by leleuj** [comment]
I'm not sure to understand your point here. Is this an example provided or a behavior you want to change in pac4j?

**Comment by jehugaleahsa** [comment]
This is a suggested behavior change. In the `SecurityHandler`, the call to `perform` should be automatically updating the HTTP response within the passed in context. By the time `perform` exits, the response should be properly configured and you shouldn't need to do any additional work, besides prevent the actual handler from firing.

Here's what my `ConfigFactory`'s build method looks like:

```java
    @Override
    public Config build(Object... objects) {
        FormAuthenticator formAuthenticator = formAuthenticatorProvider.get();
        FormClient formClient = new FormClient("/login", formAuthenticator);

        Clients clients = new Clients("/login/callback", formClient);
        clients.setAjaxRequestResolver(new DefaultAjaxRequestResolver());

        return new Config(clients);
    }
```

If you set the `setAjaxRequestResolver`, there is code in the `perform` method that will automatically detect when a request is coming from AJAX vs a form post. PAC4J will return a 401 for AJAX requests (this is currently the only behavior for javalin-pac4j); otherwise, for form posts, it will redirect the user back to the login screen with a query string containing the username and the error information. The `SecurityHandler` throwing an `UnauthorizedResponse` circumvents the second workflow and a 401 is always returned.

The Javalin [docs](https://javalin.io/documentation#handlers) are not exactly clear on how to prevent the Endpoint handler from firing after the before handler (i.e., how to treat a before handler as filter). My guess is you throw an exception, like you're doing here. Is it a matter of throwing a different [response](https://javalin.io/documentation#default-responses) exception? There doesn't appear to be a "do nothing" response type. Another idea is just create a raw `HttpResponseException` with whatever status code the response has after the call to `perform`.

Sorry for this being so long.

**Comment by leleuj** [comment]
OK. I think I get it.

If the result is not `AUTH_GRANTED`,  we must block the access as the user is not granted thus the thrown `UnauthorizedResponse`.
That said, you are right, we always block with a 401 response while it can be a 401 or a 302 or even other responses.

The signature of the method is `void` so I don't think: `return result != null;` really makes sense.
I think the right option is to throw a raw `HttpResponseException` and let pac4j set the status code and other information.

Would you mind testing that option and submitting a PR? Thx

**Comment by jehugaleahsa** [comment]
Sounds good.

**Comment by jehugaleahsa** [comment]
Just a heads up, I was trying to get this to compile and it wasn't finding the SAML dependencies, even after adding the snapshots repository to my .m2 (I saw it was already in the pom.xml later). I saw that 4.2.0 exists in the Shibboleth repository, so I added that and it got me working.

**Comment by jehugaleahsa** [comment]
@tipsy I am taking a dive into this code. One thing that would clean it up is if we could just throw `HttpResponseException` for any unauthorized action. For `RedirectResponse`, is there any way to specify the `Location` header or otherwise make it available? Should `Location` be an optional property of the `RedirectResponse` class?

**Comment by tipsy** [comment]
> @jehugaleahsa For RedirectResponse, is there any way to specify the Location header or otherwise make it available? Should Location be an optional property of the RedirectResponse class?

Maybe - redirects are a bit tricky, we were considering if we should rework them for Javalin 5. If possible, I think it's better if you use `ctx.redirect()` for now.

**Comment by jehugaleahsa** [comment]
@tipsy I feel like I keep running in circles. Minimally, I know I need to throw an exception to exit the before handler to prevent the HTTP handler from firing.

I have been looking at the `JEEHttpActionAdapter` class, which ships with pac4j. It is just setting the status code, response, headers, etc. on the underlying `HttpServletResponse`. This would work as-is if there was a way to tell the HTTP handler not to fire then.

I keep wondering 1) will anything weird happen if code directly modifies the underlying `HttpServletResponse`? and 2) could we introduce a new exception that says "skip the rest of the before handlers and HTTP handler but don't modify the response object"? If the answer to #1 is "yes", I think just calling the appropriate `ctx.result`/`ctx.status`/etc. makes sense. I don't know if #2 makes sense or not, so I wanted to get your thoughts: Is there a another way to skip the HTTP handler?

**Comment by tipsy** [comment]
> Minimally, I know I need to throw an exception to exit the before handler to prevent the HTTP handler from firing.

This is correct.

> 1) will anything weird happen if code directly modifies the underlying HttpServletResponse?

No, that should be fine. 

> 2) could we introduce a new exception that says "skip the rest of the before handlers and HTTP handler but don't modify the response object"?

This is what we're looking at for Javalin 5. The issue with making it an exception is that exception mapping can be disabled by the user. You can easily make your own `Pac4jRedirectException` and add an exception mapper for this (`app.exception(Pac4jRedirectException.class, ...)`), which should let you do whatever you want quite easily. The problem is that whoever is using your plugin could then call `app.exception(Pac4jRedirectException.class, someOtherHandler)` again, which would break your plugin. We could also introduce a boolean to prevent overwriting.

**Comment by leleuj** [comment]
@tipsy is your solution working with the Javalin version 3.13.13?

Would it be easier if we upgrade to Javalin 4.6.4? In any case, should we upgrade to Javalin 4?

**Comment by tipsy** [comment]
> @tipsy is your solution working with the Javalin version 3.13.13?

Yes, it should.

> Would it be easier if we upgrade to Javalin 4.6.4? In any case, should we upgrade to Javalin 4?

This part would be the same, and it will be the same in Javalin 5, which will be out in a few months.

Javalin 5 might include something to solve this in a different way, but this approach will still work.

**Comment by leleuj** [comment]
Excellent! OK for this solution then.

**Comment by leleuj** [comment]
@jehugaleahsa Any progress on this?

**Comment by leleuj** [comment]
Closed due to inactivity

**Comment by jehugaleahsa** [comment]
Sorry for not responding. I decided the work I had done introduced too many risks. I want to wait until Javalin 5 comes out to see if there's a better mechanism for escaping a request. I was vetting the use of javalin for a project at work. I was able to show authentication via pac4j was still relatively easy without using this library exclusively, but unfortunately lost that battle, and we're using Jersey. :( Will hopefully have another reason to get more involved in the upcoming months.

**Comment by tipsy** [comment]
> I want to wait until Javalin 5 comes out to see if there's a better mechanism for escaping a request.

It will most likely be released tomorrow. The new approach is 

```kotlin
app.before {
    (it as JavalinServletContext).tasks.clear()
}
```

This will clear any task that Javalin had remaining (any http handler, any after handler).

**Comment by leleuj** [comment]
I just upgraded the master branch (now 6.0.0-SNAPSHOT) to Javalin v5.

**Comment by jehugaleahsa** [comment]
Nice! I'll start reading the javalin 5 docs, see what's changed. Yay!

**Comment by jehugaleahsa** [comment]
Just an update: I started upgrading my [Javalin MVC](https://github.com/jehugaleahsa/javalin-mvc/tree/5.0.0) project to Javalin 5.x. It's going well - I got all my unit tests passing last night. There's a couple breaking changes I have been waiting for a major release to do, and then I need to figure out how OpenAPI works with the latest library.

After I deploy, I have a "starter app" project I will update to 5.x as well. That's where I am using this project, so I will have a place to experiment again.

**Comment by jehugaleahsa** [comment]
Another update: I finished upgrading Javalin MVC 5.0.0. I created a new repository: [javalin_mvc_starter](https://github.com/jehugaleahsa/javalin-mvc-starter); however, to get it to work, I ended up needing to reference pac4j-javalin 5.0.2-SNAPSHOT, as you can see here: https://github.com/jehugaleahsa/javalin-mvc-starter/blob/main/web/pom.xml#L65

When I tried version 5.0.1, it was still referencing the `javax.*` classes instead of `jakarta.*`.

**Comment by jehugaleahsa** [comment]
Now that I have my starter project using Javalin 5.x, I was trying to work on this again this morning. I noticed a couple things:
1. When I first clone the repository, calling `mvn install` fails. It fails due to the spotbugs plugin.
   * What is the `mvn` command to just install locally for my testing?
   * I disabled the spotbugs plugin, and then I got javadoc failures
2. The new `SecurityHandler` logic would look like this:
```java
@Override
    public void handle(@NotNull Context javalinCtx) {
        final SessionStore bestSessionStore = FindBest.sessionStore(null, config, JEESessionStore.INSTANCE);
        final HttpActionAdapter bestAdapter = FindBest.httpActionAdapter(null, config, JavalinHttpActionAdapter.INSTANCE);
        final SecurityLogic bestLogic = FindBest.securityLogic(null, config, DefaultSecurityLogic.INSTANCE);

        JavalinWebContext context = new JavalinWebContext(javalinCtx);
        Object result = bestLogic.perform(
            context,
            bestSessionStore,
            this.config,
            (ctx, sessionStore, profiles, parameters) -> AUTH_GRANTED,
            bestAdapter,
            this.clients,
            this.authorizers,
            this.matchers
        );
        if (result != AUTH_GRANTED) {
            ((JavalinServletContext) javalinCtx).getTasks().clear(); // Used to throw UnauthorizedResponse
        }
    }
```

It's hard to test this because `JavalinServletContext` is effectively marked `final`, so Mockito complains if I try to mock it. If I don't mock it, I get a `ClassCastException` when the code inside the `if` statement runs during test.

Thoughts on a good approach to move forward?

**Comment by leleuj** [comment]
I just synced with the master and successfully build the project with: `mvn clean install`. What Maven version do you use? What JDK?

I set the master to version `6.0.0-SNAPSHOT` (it should not be 5.0.2-SNAPSHOT) as it's a major upgrade to Javalin v5 and pac4j-jakartaee (instead of pac4j-javaee).

I think the best would be to ask the Javalin project to remove the `final` keyword and test with the latest Javalin SNAPSHOT.
If this is not possible, I disabled the failing test and made your change so that you can manually test (with the latest 6.0.0-SNAPSHOT). Can you check if it works? Thanks

**Comment by leleuj** [comment]
Assuming everything is OK, I just released javalin-pac4j v6.0.0.
